scanpy/preprocessing/normalization.py

```diff
@@ -0,0 +1,204 @@
+
+    Returns
+    -------
+    Returns or updates `adata` with normalized version of the original
+    `adata.X`, depending on `copy`.
+
+    Examples
+    --------
+    >>> adata = AnnData(np.array([[1, 0, 1], [3, 0, 1], [5, 6, 1]]))
+    >>> sc.pp.normalize_quantile(adata, quantile=0.7)
+    >>> print(adata.X)
+    [[1.         0.         1.        ]
+     [3.         0.         1.        ]
+     [0.71428573 0.85714287 0.14285715]]
+    """
+    if quantile < 0 or quantile > 1:
+        raise ValueError('Choose quantile between 0 and 1.')
+
+    if key_n_counts is None: key_n_counts = 'n_counts'
+
+    adata = data.copy() if copy else data
+    X = adata.X
+    gene_subset = None
+
+    if quantile < 1:
+        logg.msg('normalizing by count per cell for \
+                  genes that make up less than quantile * total count per cell', r=True)
+        X = adata.X
+
+        counts_per_cell = counts_per_cell if counts_per_cell is not None else X.sum(1)
+        counts_per_cell = np.ravel(counts_per_cell)
+
+        gene_subset = (X>counts_per_cell[:, None]*quantile).sum(0)
+        gene_subset = (np.ravel(gene_subset) == 0)
+    else:
+        logg.msg('normalizing by total count per cell', r=True)
+
+    if counts_per_cell is None or quantile < 1:
+        X = X if gene_subset is None else adata[:, gene_subset].X
+        counts_per_cell = X.sum(1)
+        counts_per_cell = np.ravel(counts_per_cell)
+    del X
+    del gene_subset
+
+    adata.obs[key_n_counts] = counts_per_cell
+    cell_subset = counts_per_cell >= min_counts
+    adata._inplace_subset_obs(cell_subset)
+    counts_per_cell = counts_per_cell[cell_subset]
+
+    if use_rep == 'after':
+        after = counts_per_cell_after
+    elif use_rep == 'X':
+        after = np.median(counts_per_cell)
+    elif use_rep is None:
+        after = None
+    else: raise ValueError('use_rep should be "after", "X" or None')
+
+    _normalize_data(adata.X, counts_per_cell, counts_per_cell_after)
+
+    layers = adata.layers.keys() if layers == 'all' else layers
+    for layer in layers:
+        L = adata.layers[layer]
+        counts = np.ravel(L.sum(1))
+        _normalize_data(L, counts, after)
+
+    logg.msg('    finished', t=True, end=': ')
+    logg.msg('normalized adata.X and added', no_indent=True)
+    logg.msg('    \'{}\', counts per cell before normalization (adata.obs)'
+        .format(key_n_counts))
+
+    return adata if copy else None
+
+def normalize_total(data, counts_per_cell_after=None, counts_per_cell=None,
+                    key_n_counts=None, copy=False, layers=[], use_rep=None,
+                    min_counts=1):
+    """Normalize total counts per cell.
+
+    Normalize each cell by total counts over all genes, so that every cell has
+    the same total count after normalization.
+
+    Similar functions are used, for example, by Seurat [Satija15]_, Cell Ranger
+    [Zheng17]_ or SPRING [Weinreb17]_.
+
+    Parameters
```

Comment:
We don't want to replicate the docstring here. Please use the decorator `@doc_params` that is used in wide parts of Scanpy for avoiding replicated docstrings. Both `normalize_total` and `normalize_counts` should be based on the same docstring. Take a look at https://github.com/theislab/scanpy/blob/master/scanpy/plotting/tools/scatterplots.py, for instance.